Add LIST_CONCAT and LIST_APPEND functions

lnkuiper opened the pull request:
This PR implements `list_concat(listA, listB)` that appends lists to each other.
```sql
SELECT list_concat([1, 2], [3, 4])
--- [1, 2, 3, 4]
```
It has 3 aliases: `list_cat`, `array_concat` and `array_cat`.

I've added the macro `list_append(list, element)` that appends a value to a list (`list_append(l, e)` gets rewritten to `list_append(l, list_value(e))`. This has one alias: `array_append`.

@szarnyasg

Commit:
add list_concat function

Commit:
undo wrong change

Mytherin commented on src/function/scalar/list/list_concat.cpp:
```diff
@@ -0,0 +1,132 @@
+
+	auto &lhs = arguments[0]->return_type;
+	auto &rhs = arguments[1]->return_type;
+	if (lhs.id() == LogicalTypeId::SQLNULL) {
```
Perhaps here we can copy the bind from `list_value.cpp` (`list_value.cpp:37`). That way we should be able to add implicit casts in case of slightly different types (e.g. `SMALLINT` and `INT`, or similar).

Mytherin commented on test/sql/function/list/list_concat.test:
```diff
@@ -0,0 +1,113 @@
+query II
+SELECT i, list_append(list_concat(j, k), i) FROM lists
+----
+0	[0, 4, 8, 12, 0, 4, 8, 12, 0]
```
Could we add some tests with lists of different types (e.g. `list_concat([1::tinyint], [2::int])`, or more subtly `list_concat([0.1], [10])`

Mytherin commented on src/function/scalar/list/list_concat.cpp:
```diff
@@ -0,0 +1,132 @@
+	ScalarFunction lfun({LogicalType::LIST(LogicalType::ANY), LogicalType::LIST(LogicalType::ANY)},
+	                    LogicalType::LIST(LogicalType::ANY), ListConcatFunction, false, ListConcatBind, nullptr,
+	                    ListConcatStats);
+	set.AddFunction({"list_concat", "list_cat", "array_concat", "array_cat"}, lfun);
```
In Python we can also use the + operator to concatenate lists, perhaps also add that overload? This will need to be added to `arithmetic.cpp:310`.

Mytherin commented on src/function/scalar/list/list_concat.cpp:
```diff
@@ -0,0 +1,132 @@
+	                    LogicalType::LIST(LogicalType::ANY), ListConcatFunction, false, ListConcatBind, nullptr,
+	                    ListConcatStats);
+	set.AddFunction({"list_concat", "list_cat", "array_concat", "array_cat"}, lfun);
+}
```
Postgres also overloads [`||` for lists ](https://www.postgresql.org/docs/9.1/functions-array.html), perhaps we should allow the same.

Mytherin commented on test/sql/function/list/list_concat.test:
```diff
@@ -0,0 +1,113 @@
+SELECT i, list_append(list_concat(j, k), i) FROM lists
+----
+0	[0, 4, 8, 12, 0, 4, 8, 12, 0]
+1	[1, 5, 9, 13, 1, 5, 9, 13, 1]
```
Could we add a test with nested lists (lists in lists in lists)?

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Some comments:

Commit:
PR feedback

lnkuiper commented:
@Mytherin thanks for the feedback. I've implemented it.

I've also taken a look at Postgres behaviour, and changed NULL behaviour. `list_concat(NULL, [3, 4])` now yields `[3, 4]`.

Mytherin commented on src/function/scalar/operators/arithmetic.cpp:
```diff
@@ -328,6 +334,8 @@ void AddFun::RegisterFunction(BuiltinFunctions &set) {
 	functions.AddFunction(GetFunction(LogicalType::TIMESTAMP, LogicalType::INTERVAL));
 	functions.AddFunction(GetFunction(LogicalType::INTERVAL, LogicalType::TIMESTAMP));
+	// we can add lists together
+	functions.AddFunction(GetFunction(LogicalType::LIST(LogicalType::ANY), LogicalType::LIST(LogicalType::ANY)));
```
This could just be `functions.AddFunction(ListConcatFun::GetFunction())`, no?

lnkuiper replied:
Yes, I will change it.

Mytherin reviewed (commented):
Thanks for the updates! One more comment then this is good to go.

Commit:
simplify list + operator function registration

Mytherin commented on src/function/scalar/operators/arithmetic.cpp:
```diff
@@ -285,6 +286,11 @@ ScalarFunction AddFun::GetFunction(const LogicalType &left_type, const LogicalTy
 			                      ScalarFunction::BinaryFunction<interval_t, timestamp_t, timestamp_t, AddOperator>);
 		}
 		break;
+	case LogicalTypeId::LIST:
```
This can be removed now I believe

lnkuiper replied:
You're right

Mytherin commented on src/catalog/default/default_functions.cpp:
```diff
@@ -86,6 +86,8 @@ static DefaultMacro internal_macros[] = {
 	{DEFAULT_SCHEMA, "roundbankers", {"x", "n", nullptr}, "round_even(x, n)"},
 	{DEFAULT_SCHEMA, "nullif", {"a", "b", nullptr}, "CASE WHEN a=b THEN NULL ELSE a END"},
+	{DEFAULT_SCHEMA, "list_append", {"l", "e", nullptr}, "list_concat(l, list_value(e))"},
+	{DEFAULT_SCHEMA, "array_append", {"arr", "el", nullptr}, "list_append(arr, el)"},
```
While we're at it we can add array_prepend here as well right?

lnkuiper replied:
Great idea. Added list_prepend and array_prepend

Commit:
add array prepend as well

Mytherin reviewed (approved):
Thanks for the updates! Looks great now.
